File: c/tree/binary-tree-preorder-traversal[1].c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
void recursive(struct TreeNode* root, int **array, int *idx, int *arr_size)
{
    if(!root)
        return ;


    if(*idx >= *arr_size){
        *arr_size <<= 1;
        *array = realloc(*array, sizeof(int) * (*arr_size));
    }

    int *arr = *array;
    arr[(*idx)++] = root->val;
    recursive(root->left,&arr,idx,arr_size);
    recursive(root->right,&arr,idx,arr_size);
}

int* preorderTraversal(struct TreeNode* root, int* returnSize) {
    int *arr = NULL;
    int arr_size = 32;
    *returnSize = 0;

    arr = malloc(sizeof(int) * arr_size);
    recursive(root,&arr,returnSize,&arr_size);

    return arr;
}

File: python/tree/binary-tree-preorder-traversal[1].py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        st = []
        arr = []
        if root:
            st.append(root)

        while st:
            node = st.pop()
            arr.append(node.val)
            if node.right:
                st.append(node.right)
            if node.left:
                st.append(node.left)

        return arr
